### ISSUE TITLE
FedSTAR

### ISSUE THREAD

**Issue by jafermarq** [adap/flower#2042]
# FedSTAR

* **Title:** Federated Self-training for Semi-supervised Audio Recognition
* **Venue:** ACM TEC 2022
* **Link to paper:** https://dl.acm.org/doi/10.1145/3520128

## Do you want to work on this baseline? 

> 🌻 Check everything about the `Summer of Reproducibility` on [`flower.dev/summer`](https://flower.dev/summer/)

> All available baselines are listed in the [Summer of Reproducibility Dashboard](https://github.com/orgs/adap/projects/6/) and also in the [GitHub Issues with the `summer-of-reproducibility` label](https://github.com/adap/flower/labels/summer-of-reproducibility). The content is the same.

> 📝 It is advised to complete these steps before your start working on your code. But if you can't wait to implement your baseline with Flower (we totally understand it 😄), please ensure you follow the steps on [how to contribute a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline).

What follows are the steps 1 & 2 in the Summer of Reproducibility instructions.

### 1. Join the Summer of Reproducibility program
- [x] Join the [Flower Slack](https://flower.dev/join-slack/) and say _"hi! 👋"_ in the channel `#summer-of-reproducibility`.
- [x] Pick a baseline from our curated list **<---------------------------------------- [you are doing this now]**
### 2. Define the scope of your contribution
- [x] **What are you going to reproduce?** Add a comment to your issue and tell us about your plan regarding this baseline: what experiments from the paper are you reproducing?, for which datasets ? the more details you provide us with the better !
- [x] **Check if you are eligible for a reward**. 
    > As we have to comply with US/EU regulations, we have checked that individual contributors based on these countries or territories are eligible: Australia, Austria, Belgium, Bulgaria, Canada, Croatia, Cyprus, Czech Republic, Denmark, Estonia, Finland, France, Germany, Gibraltar, Greece, Hong Kong SAR China, Hungary, India, Ireland, Italy, Japan, Latvia, Liechtenstein, Lithuania, Luxembourg, Malta, Mexico, Netherlands, New Zealand, Norway, Poland, Portugal, Romania, Singapore, Slovakia, Slovenia, South Korea, Spain, Sweden, Switzerland, Thailand, United Arab emirates, United Kingdom United States.

    If where you are based is not on the list, please send us an email (`summer@flower.dev`) letting us know a bit about yourself (where are you currently based?, are you a university student? do you work at a public institution?). Please tell us the baselines you are interested in implementing (i.e. tell us your GitHub issue if you have crated one). We will reach back to you.
- [x] We will discuss with you about your contribution plan, if it sounds like a substantial enough contribution according to the Summer of Reproducibility rules (check our website [flower.dev/summer](https://flower.dev/summer/)), you'll get the OK from us to start working on your baseline!

## What happens next?

- [x] This item will be moved to the `In Progress` stage by a member of the Flower Team.

- [x] Follow the [instructions for creating a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline) which will guide you through the process step-by-step.

## Is something wrong or not clear ?
* Ask a question directly in your issue.
* Reach out to us via the [Flower Slack](https://flower.dev/join-slack/) and ask your question in the #summer-of-reproducibility channel
* Check all the details (including FAQ) in the Summer of Reproducibility website: flower.dev/summer

**Comment by Raj-Parekh24** [comment]
Hi @jafermarq!

I would like to implement the FedSTAR baseline in the flower framework

This is the contribution plan I'd like to propose:

-> I will be working on the speech commands task as it consists of 12 classes and the crucial results in the paper are discussed 
     using this task.
-> Implement a centralized model to set as a benchmark model to compare results with federated learning architecture as 
     shown in Table 3.
-> Implementation of the teacher-student model on the client side for semi-supervised learning.
-> Varying the client's resource usage to check the performance of FedStar shown in Fig 3 because all clients will not be able to 
     give there 100% computing power.
-> Varying the number of clients participating in local training procedures to reproduce results shown in Fig 5.
-> Checking the performance of FedStar by varying the number of classes available for training at each client side to reproduce 
     results are shown in Table 5.
-> Check the global loss values over the communication with the pre-trained model and randomly initialized model to check the 
     convergence shown in Fig 6a, Fig 6b.

**Comment by jafermarq** [comment]
hi @Raj-Parekh24, it's great to hear you want to reproduce the results in `FedSTAR`. Just so be sure I got it right, your plan is to reproduce:
* The top row in Table 3 (centralised)
* Figure 5
* All results in Table 5
* Figure 6 a & b

I agree having the centralised baseline (as in Table 3) and the results in Table 5 is good. But we encourage all contributors in the _Summer of Reproducibility_ to include at least two datasets. Would it make sense to choose another of the datasets (maybe `AmbientContext?`) and focus in reproducing Table 4 for both SpeechCommands and the second dataset? In my view Table 4 is the main result of the paper. I would leave those in Fig5 and 6a,b as an extension?

**Comment by Raj-Parekh24** [comment]
Hey @jafermarq ,

Sounds good 

Updated tasks:-

Table 3,4 Experiments with the task of Ambient Context, and Speech Commands
Table 5, Figure 5, 6a, and 6b based on Speech Commands for additional verification.

Thanks

**Comment by jafermarq** [comment]
@Raj-Parekh24, could you also let me know if you'd be eligible (once all other steps in the program are completed) for a reward? Essentially, you need to check if where you are based it's in the list shown after the second point in Step 2. And to double check, are you sure you'll have the time to implement all those experiments? it doesn't sound like little work :)

**Comment by jafermarq** [comment]
Perfect! many thanks for putting the effort to revisit your plan and re-read the paper. I have now ✅ everything in Step 1 & 2, added you as the _assignee_ to this issue, and labelled this baseline as _In Progress_. You can find instructions on how to start with the code by following the link shown in the _What happens Next?_ section in above in the description of this issue. If you have any doubts or suggestions, please reach out to me and the other contributors in the _Summer of Reproducibility_ via the Flower Slack.

Looking forward to seeing your `FedSTAR` implementation in action! It's nice to see some speech baselines after so many doing vision tasks :)

**Comment by Raj-Parekh24** [comment]
Thanks a lot for all the help @jafermarq ,

Excited to work on this project.

**Comment by jafermarq** [comment]
Hi @Raj-Parekh24,

This is just a gentle reminder that the Flower *Summer of Reproducibility* is ending at the end of the month. With just a little more than 3 weeks to go, we are excited to see quite a few baselines well ahead in the process with [their respective PRs close to ready.](https://github.com/adap/flower/pulls?q=is%3Apr+is%3Aopen+label%3Asummer-of-reproducibility) If your PR is already on the list, great !! Please make sure the PR is linked to this issue (you just need to copy the URL of this issue somewhere in the main message of your PR). Ping me when you'd like me to take a look.

Also, make sure you keep an eye:eyes: on the **#summer-of-reproducibility** channel in the Flower Slack. I’ll announce very soon a new (the third!) round of 1:1 ask-me-anything sessions to help _Summer of Reproducibility_ contributors like yourself to meet the deadline. Please consider booking a time slot if you want to chat with me about your baseline, potential issues you have making your code run, how to open a PR, doubts about what to include in your readme, how to use Hydra configs more effective, etc … all questions are welcome!!

**Comment by jafermarq** [comment]
Congratulations @Raj-Parekh24 your baseline has been merged. We'll follow up with you soon to complete your _Summer of Reproducibility_ journey 😄 !

**Comment by Raj-Parekh24** [comment]
Thanks a lot @jafermarq!
